app/V1Module/presenters/AssignmentSolutionsPresenter.php

```diff
@@ -240,27 +263,31 @@ public function actionDownloadSolutionArchive(string $id) {
     $this->sendResponse(new ZipFilesResponse($files, "solution-{$id}.zip"));
   }
 
+  public function checkDownloadResultArchive(string $id) {
+    $submission = $this->assignmentSolutionSubmissions->findOrThrow($id);
+    if (!$this->assignmentSolutionAcl->canDownloadResultArchive($submission->getAssignmentSolution())) {
+      throw new ForbiddenRequestException("You cannot access result archive for this submission");
+    }
+
+    $this->evaluationLoadingHelper->loadEvaluation($submission);
+
+    if (!$submission->hasEvaluation()) {
```

Comment:
This in my opinion is not authorization check and should be in action method. User which have permissions to see button for downloading archive should see it even if it is not ready yet...

Reply:
Yeah, I'll probably give you this one.